### PR TITLE
feat(core): rules setup

### PR DESCRIPTION
**Pull request by YvesRijckaert**
- Move "np" from dependencies to devDependencies
- Move stylelint to peerDependencies
- Create a rules folder and move the "core" rules from index.js into rules/core.js. The index file exposse the core rules
- Sort and document the rules in the core section (with links to doc)
- Prepare a separate rules file for "styled-components" and "tailwind"
- Husky /Lint-staged configuration
- Remove babel-eslint
- ESLint all *.js files
- Add some keywords to the package.json
- Put yarn into the devDependencies (since we using yarn instead of npm)

### CODE REVIEW

File: rules/core.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
+module.exports = {
+  extends: ["stylelint-config-standard", "stylelint-config-rational-order"],
+  plugins: ["stylelint-order", "stylelint-config-rational-order/plugin"],
+  rules: {
+    /* Specify the alphabetical order of properties within declaration blocks. */
+    /* https://github.com/hudochenkov/stylelint-order/tree/master/rules/properties-alphabetical-order */
+    "order/properties-alphabetical-order": null,
+    /* Specify the order of properties within declaration blocks. */
```
Comment (schorfES) [inline rules/core.js:8]: Can we separate each comment + rule with a new line for a better readability?

```
// Specify the alphabetical order of properties within declaration blocks.
// https://github.com/hudochenkov/stylelint-order/tree/master/rules/properties-alphabetical-order
"order/properties-alphabetical-order": null,

// Specify the order of properties within declaration blocks.
// https://github.com/hudochenkov/stylelint-order/tree/master/rules/properties-order
"order/properties-order": [],
```

(I've also change the comment style. Doesn't make sense to me to use multiline-comments in a single line. Break the line between 80-96 characters)

File: rules/styled-components.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    "stylelint-config-styled-components",
+  ],
+  plugins: [
+    "stylelint-order",
+    "stylelint-config-rational-order/plugin",
+  ],
+  rules: {
+    /* Specify the alphabetical order of properties within declaration blocks. */
+    /* https://github.com/hudochenkov/stylelint-order/tree/master/rules/properties-alphabetical-order */
+    "order/properties-alphabetical-order": null,
```
Comment (schorfES) [inline rules/styled-components.js:17]: Please do not re-introduce rules from core. This file should only extend the core configuration. The usage in the stylelint configuration should look like:

```
{
  "extends": [
    "@edenspiekermann/stylelint-config",
    "@edenspiekermann/stylelint-config/styled-components"
  ]
}
```

File: rules/tailwind.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  ],
+  plugins: [
+    "stylelint-scss",
+    "stylelint-order",
+    "stylelint-config-rational-order/plugin",
+  ],
+  rules: {
+    /* Specify the alphabetical order of properties within declaration blocks. */
+    /* https://github.com/hudochenkov/stylelint-order/tree/master/rules/properties-alphabetical-order */
+    "order/properties-alphabetical-order": null,
```
Comment (schorfES) [inline rules/tailwind.js:14]: Please do not re-introduce rules from core. This file should only extend the core configuration. The usage in the stylelint configuration should look like:

```
{
  "extends": [
    "@edenspiekermann/stylelint-config",
    "@edenspiekermann/stylelint-config/tailwind"
  ]
}
```

or

```
{
  "extends": [
    "@edenspiekermann/stylelint-config",
    "@edenspiekermann/stylelint-config/sass",
    "@edenspiekermann/stylelint-config/tailwind"
  ]
}
```

File: rules/core.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    /* https://stylelint.io/user-guide/rules/media-feature-name-case */
+    "media-feature-name-case": "lower",
+    /* Specify lowercase or uppercase for keywords values. */
+    /* https://stylelint.io/user-guide/rules/value-keyword-case */
+    "value-keyword-case": "lower",
+    /* Disallow extra semicolons. */
+    /* https://stylelint.io/user-guide/rules/no-extra-semicolons */
+    "no-extra-semicolons": true,
+  },
+  ignoreFiles: ["./.cache/**/*", "./node_modules/**/*", "./public/**/*"],
```
Comment (schorfES) [inline rules/core.js:98]: I would suggest to only apply this to the `node_modules`. The rest is a project specific configuration. Please double-check if stylelint needs to ignore `node_modules` or ignores them by default...

File: package.json [hunk 5]
```diff
@@ -3,8 +3,16 @@
   "version": "0.0.1",
   "description": "A shareable stylelint config for ESPI projects",
   "main": "index.js",
+  "keywords": [
+    "stylelint",
+    "stylelint config",
+    "prettier",
```
Comment (schorfES) [inline package.json:9]: Prettier?

**Review by schorfES — CHANGES_REQUESTED** [review]
I've added some comments. Please also update the README.md according the different rulesets.

File: rules/sass.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+module.exports = {
+  extends: [
+    '@edenspiekermann/stylelint-config',
```
Comment (schorfES) [inline rules/sass.js:3]: Why self-extending?

File: rules/styled-components.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
+module.exports = {
+  processors: [
+    'stylelint-processor-styled-components',
+    'stylelint-config-rational-order',
+  ],
+  extends: [
+    '@edenspiekermann/stylelint-config',
```
Comment (schorfES) [inline rules/styled-components.js:7]: Why self-extending?

File: rules/tailwind.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
+module.exports = {
+  extends: [
+    '@edenspiekermann/stylelint-config',
```
Comment (schorfES) [inline rules/tailwind.js:3]: Why self-extending?

File: rules/styled-components.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    '@edenspiekermann/stylelint-config/rules/sass',
+    '@edenspiekermann/stylelint-config/rules/styled-components',
+  ],
+  plugins: ['stylelint-order', 'stylelint-config-rational-order/plugin'],
+  rules: {
+    // Specify the alphabetical order of properties within declaration blocks.
+    // https://github.com/hudochenkov/stylelint-order/tree/master/rules/properties-alphabetical-order
+    'order/properties-alphabetical-order': null,
```
Comment (schorfES) [inline rules/styled-components.js:15]: No need to duplicate rules from an other rules set that is more generic.

**Review by schorfES — CHANGES_REQUESTED** [review]
There are a lot duplicate rules. Each rule should be placed in the most generic rules-set file. If you want to configure your project for example to use with styled-components, you need to put...

```
extends: [
  '@edenspiekermann/stylelint-config',
  '@edenspiekermann/stylelint-config/rules/styled-components',
]
```

...into you config file. There is no need to replicate rules from `@edenspiekermann/stylelint-config/rules/core` into `@edenspiekermann/stylelint-config/rules/styled-components`.

File: rules/core.js [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+       6. Misc
+    */
+    // https://www.npmjs.com/package/stylelint-config-rational-order
+    'plugin/rational-order': true,
+
+    // Require or disallow an empty line before the closing brace of blocks.
+    // https://stylelint.io/user-guide/rules/block-closing-brace-empty-line-before
+    'block-closing-brace-empty-line-before': null,
+
+    // There must always be an empty line before at-rules.
```
Comment (YvesRijckaert) [inline rules/core.js:28]: wrong documentation link

File: rules/core.js [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+    // Specify whether or not quotation marks should be used around font family names.
+    // Quotes are recommended in the spec with "font family names
+    // that contain white space, digits, or punctuation characters other than hyphens".
+    // https://stylelint.io/user-guide/rules/font-family-name-quotes
+    'font-family-name-quotes': 'always-where-recommended',
+
+    // Require (where possible) or disallow named colors.
+    // https://stylelint.io/user-guide/rules/color-named
+    'color-named': 'always-where-possible',
```
Comment (YvesRijckaert) [inline rules/core.js:52]: back to only hex

Comment (YvesRijckaert) [inline rules/core.js:52]: also include a rule for color-hex-length (use long)

File: rules/core.js [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+    // https://stylelint.io/user-guide/rules/string-quotes
+    'string-quotes': 'single',
+
+    // Disallow duplicate selectors within a stylesheet.
+    // https://stylelint.io/user-guide/rules/no-duplicate-selectors
+    'no-duplicate-selectors': true,
+
+    // Limit the number of decimal places allowed in numbers.
+    // https://stylelint.io/user-guide/rules/number-max-precision
+    'number-max-precision': 2,
```
Comment (YvesRijckaert) [inline rules/core.js:76]: use 3

File: rules/core.js [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+
+    // Disallow trailing zeros in numbers.
+    // https://stylelint.io/user-guide/rules/number-no-trailing-zeros
+    'number-no-trailing-zeros': true,
+
+    // Require or disallow a leading zero for fractional numbers less than 1.
+    // https://stylelint.io/user-guide/rules/number-leading-zero
+    'number-leading-zero': 'never',
```
Comment (YvesRijckaert) [inline rules/core.js:84]: always

File: rules/core.js [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+    // https://stylelint.io/user-guide/rules/shorthand-property-no-redundant-values
+    'shorthand-property-no-redundant-values': true,
+
+    // Disallow qualifying a selector by type. Allow attribute selectors qualified by type.
+    // https://stylelint.io/user-guide/rules/selector-no-qualifying-type
+    'selector-no-qualifying-type': [true, { ignore: 'attribute' }],
+
+    // Require a newline or disallow whitespace after the commas of selector lists.
+    // https://stylelint.io/user-guide/rules/selector-list-comma-newline-after
+    'selector-list-comma-newline-after': null,
```
Comment (YvesRijckaert) [inline rules/core.js:96]: always-multi-line

File: rules/core.js [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+
+    // Limit the number of ID selectors in a selector
+    // https://stylelint.io/user-guide/rules/selector-max-id
+    'selector-max-id': 0,
+
+    // Limit the number of universal selectors in a selector.
+    // https://stylelint.io/user-guide/rules/selector-max-universal
+    'selector-max-universal': 0,
```
Comment (YvesRijckaert) [inline rules/core.js:104]: use 1

File: rules/core.js [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+
+    // Disallow vendor prefixes for selectors.
+    // https://stylelint.io/user-guide/rules/selector-no-vendor-prefix
+    'selector-no-vendor-prefix': true,
```
Comment (YvesRijckaert) [inline rules/core.js:108]: test if it works with scrollbar

File: rules/sass.js [hunk 17]
```diff
@@ -0,0 +1,9 @@
...
+    '@edenspiekermann/stylelint-config/rules/sass',
+  ],
+  plugins: [
+    'stylelint-scss',
+  ],
```
Comment (YvesRijckaert) [inline rules/sass.js:8]: rule for mixins on top of the declaration

File: rules/styled-components.js [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+    'stylelint-processor-styled-components',
+    'stylelint-config-rational-order',
+  ],
+  extends: [
+    '@edenspiekermann/stylelint-config',
+    '@edenspiekermann/stylelint-config/rules/sass',
+    '@edenspiekermann/stylelint-config/rules/styled-components',
+  ],
+  plugins: ['stylelint-order', 'stylelint-config-rational-order/plugin'],
+  rules: {
```
Comment (YvesRijckaert) [inline rules/styled-components.js:12]: remove duplicates from here that we already defined in core

File: rules/styled-components.js [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+    'max-nesting-depth': 3,
+
+    // Require a single space or disallow whitespace on the inside of the parentheses of functions.
+    // https://stylelint.io/user-guide/rules/function-parentheses-space-inside
+    'function-parentheses-space-inside': null,
+
+    // Specify a pattern for class selectors.
+    // https://stylelint.io/user-guide/rules/selector-class-pattern
+    'selector-class-pattern':
+      '^(?:(?:o|c|u|t|s|is|has|_|js|qa)-)?[a-zA-Z0-9]+(?:-[a-zA-Z0-9]+)*(?:__[a-zA-Z0-9]+(?:-[a-zA-Z0-9]+)*)?(?:--[a-zA-Z0-9]+(?:-[a-zA-Z0-9]+)*)?(?:\\[.+\\])?$',
```
Comment (YvesRijckaert) [inline rules/styled-components.js:54]: remove from first question mark to third question mark

Comment (YvesRijckaert) [reply rules/styled-components.js:54]: (?:(?:o|c|u|t|s|is|has|_|js|qa)-)?

File: rules/tailwind.js [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+    '@edenspiekermann/stylelint-config/rules/sass',
+    '@edenspiekermann/stylelint-config/rules/tailwind',
+  ],
+  plugins: [
+    'stylelint-scss',
+    'stylelint-order',
+    'stylelint-config-rational-order/plugin',
+  ],
+  rules: {
```
Comment (YvesRijckaert) [inline rules/tailwind.js:12]: delete duplicates that are already defined in core

File: rules/styled-components.js [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+    // https://stylelint.io/user-guide/rules/max-nesting-depth
+    'max-nesting-depth': 3,
+
+    // Require a single space or disallow whitespace on the inside of the parentheses of functions.
+    // https://stylelint.io/user-guide/rules/function-parentheses-space-inside
+    'function-parentheses-space-inside': null,
+
+    // Specify a pattern for class selectors.
+    // https://stylelint.io/user-guide/rules/selector-class-pattern
+    'selector-class-pattern':
```
Comment (YvesRijckaert) [inline rules/styled-components.js:53]: make a separate rule for BEM

File: rules/tailwind.js [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+    // https://stylelint.io/user-guide/rules/color-hex-length
+    'color-hex-length': 'long',
+
+    // Limit the depth of nesting.
+    // https://stylelint.io/user-guide/rules/max-nesting-depth
+    'max-nesting-depth': 3,
+
+    // Require a single space or disallow whitespace on the inside of the parentheses of functions.
+    // https://stylelint.io/user-guide/rules/function-parentheses-space-inside
+    'function-parentheses-space-inside': null,
```
Comment (YvesRijckaert) [inline rules/tailwind.js:69]: change to never

Comment (YvesRijckaert) [reply rules/tailwind.js:69]: place in core

File: rules/tailwind.js [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+
+    // Specify blacklist of disallowed file extensions for partial names in @import commands.
+    // https://github.com/kristerkari/stylelint-scss/blob/master/src/rules/at-import-partial-extension-blacklist/README.md
+    'scss/at-import-partial-extension-blacklist': null,
```
Comment (YvesRijckaert) [inline rules/tailwind.js:73]: place in SASS rule file

File: rules/bem.js [hunk 24]
```diff
@@ -0,0 +1,11 @@
+module.exports = {
+  extends: [
```
Comment (schorfES) [inline rules/bem.js:2]: This rule-set still extends itself.

File: rules/sass.js [hunk 25]
```diff
@@ -0,0 +1,11 @@
+module.exports = {
+  extends: [
+    '@edenspiekermann/stylelint-config/rules/sass',
```
Comment (schorfES) [inline rules/sass.js:3]: This rule-set still extends itself.

File: rules/styled-components.js [hunk 26]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+  processors: ['stylelint-processor-styled-components'],
+  extends: [
+    '@edenspiekermann/stylelint-config/rules/styled-components',
```
Comment (schorfES) [inline rules/styled-components.js:4]: This rule-set still extends itself.

File: rules/tailwind.js [hunk 27]
```diff
@@ -0,0 +1,29 @@
+module.exports = {
+  extends: [
+    '@edenspiekermann/stylelint-config/rules/tailwind',
```
Comment (schorfES) [inline rules/tailwind.js:3]: This rule-set still extends itself.

File: rules/tailwind.js [hunk 28]
```diff
@@ -0,0 +1,29 @@
...
+  ],
+  rules: {
+    // Disallow unknown at-rules.
+    // https://stylelint.io/user-guide/rules/at-rule-no-unknown
+    'at-rule-no-unknown': [null],
```
Comment (schorfES) [inline rules/tailwind.js:8]: To disable this rule we need to set...
```suggestion
    'at-rule-no-unknown': null,
```

**Review by schorfES — CHANGES_REQUESTED** [review]
Still some small issues. Please also add a section of each ruleset to the README.md

**Review by schorfES — APPROVED** [review]

**Pull request merged**